tests/unit/core/server/testServer.js

```diff
@@ -6,57 +6,74 @@ import { reducer as reduxAsyncConnect } from 'redux-connect';
 import createSagaMiddleware from 'redux-saga';
 import NestedStatus from 'react-nested-status';
 import supertest from 'supertest';
+import config from 'config';
 
 import baseServer from 'core/server/base';
+import apiReducer from 'core/reducers/api';
+import userReducer from 'core/reducers/user';
+import userSaga from 'core/sagas/user';
 import FakeApp, { fakeAssets } from 'tests/unit/core/server/fakeApp';
+import {
+  createApiResponse,
+  createUserProfileResponse,
+  userAuthToken,
+} from 'tests/unit/helpers';
 
-describe('core/server/base', () => {
-  const _helmentCanUseDOM = Helmet.canUseDOM;
+
+describe(__filename, () => {
+  let store;
+  let mockWindow;
+  const _helmetCanUseDOM = Helmet.canUseDOM;
   const defaultStubRoutes = (
     <Router>
       <Route path="*" component={FakeApp} />
     </Router>
   );
+  const apiHost = config.get('apiHost');
 
   beforeEach(() => {
     Helmet.canUseDOM = false;
     global.webpackIsomorphicTools = {
       assets: () => fakeAssets,
     };
+    mockWindow = sinon.mock(global.window);
   });
 
   afterEach(() => {
-    Helmet.canUseDOM = _helmentCanUseDOM;
+    Helmet.canUseDOM = _helmetCanUseDOM;
     delete global.webpackIsomorphicTools;
   });
 
-  function testClient({ stubRoutes = defaultStubRoutes } = {}) {
+  function testClient({
+    stubRoutes = defaultStubRoutes,
+    reducers = { reduxAsyncConnect },
+    appSagas = null,
+  } = {}) {
     function createStoreAndSagas() {
       const sagaMiddleware = createSagaMiddleware();
-      return {
-        store: createStore(
-          combineReducers({ reduxAsyncConnect }),
-          // Do not define an initial state.
-          undefined,
-          applyMiddleware(sagaMiddleware),
-        ),
-        sagaMiddleware,
-      };
+      store = createStore(
+        combineReducers(reducers),
+        // Do not define an initial state.
+        undefined,
+        applyMiddleware(sagaMiddleware),
+      );
+
+      return { store, sagaMiddleware };
     }
 
     // eslint-disable-next-line no-empty-function
     function* fakeSaga() {}
 
     const app = baseServer(stubRoutes, createStoreAndSagas, {
-      appSagas: fakeSaga,
+      appSagas: appSagas || fakeSaga,
       appInstanceName: 'testapp',
     });
     return supertest(app);
   }
 
   describe('app', () => {
     it('varies on DNT', async () => {
-      const response = await testClient().get('/en-US/firefox/').end();
```

Comment:
Unfortunately, `.end()` must always be the final call to the `testClient()` chain. This is because `end()` will shut down the server. So, without it, every one of these tests will be spawning a server that will never get shut down. It's the most terrible interface ever written. See https://github.com/visionmedia/supertest/issues/100

I originally couldn't think of a way to do this automatically but maybe we can just use a wrapper around `testClient()` ?